### ISSUE TITLE
Add an extensibility point for ThreadLocalState

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#56027]
## 🚀 Feature

An extensibility point for [ThreadLocalState](https://github.com/pytorch/pytorch/blob/99d77c55ddb8bcfec0ec916faeaf8325a4153fe3/aten/src/ATen/ThreadLocalState.h#L14) that allows extension writers to define their own custom TLS that gets saved and propagated through the various PyTorch subsystems that use ThreadLocalState.

## Motivation

For the composable function transforms prototype (eager-mode vmap+grad), we have a TLS stack that controls how dispatching happens. For example, if the TLS stack looks like [grad, vmap, grad], we'd like to dispatch all operations on VariableType first, then run the batching rule, then dispatch on VariableType again. This TLS state should be propagated to threads, especially when autograd runs.

## Pitch

One possible implementation is to offer extension writers a `REGISTER_TLS` function and a subclass-able `ThreadLocalStateInterface` struct. `REGISTER_TLS` would accept two functions, `save_tls: () -> ThreadLocalStateInterface` and `restore_tls: (ThreadLocalStateInterface) -> void`.

## Alternatives

No clue

## Additional context

This should ideally be close to zero-cost if no one is using the extensibility feature. There should also be low unit-cost for folks who are using the extensibility feature.

**Comment by dzhulgakov** [comment]
There's already some extension mechanism for it called ThreadLocalDebugInfo: https://github.com/pytorch/pytorch/blob/99d77c55ddb8bcfec0ec916faeaf8325a4153fe3/c10/util/ThreadLocalDebugInfo.h - it's getting propagated across threads and I think has the same semantics as you want. This mechanism is already used for passing various profiling information around (so it follows torch.jit.fork, autograd execution, etc).

First, there's nothing "debug" about it, we should probably just rename it. Second, it's implemented with a bunch of shared pointers so maybe it needs to be a bit fine-tuned for perf (but maybe it's fine). cc @ilia-cher who originally added it

**Comment by ilia-cher** [comment]
> There's already some extension mechanism for it called ThreadLocalDebugInfo: https://github.com/pytorch/pytorch/blob/99d77c55ddb8bcfec0ec916faeaf8325a4153fe3/c10/util/ThreadLocalDebugInfo.h - it's getting propagated across threads and I think has the same semantics as you want. This mechanism is already used for passing various profiling information around (so it follows torch.jit.fork, autograd execution, etc).
> 
> First, there's nothing "debug" about it, we should probably just rename it. Second, it's implemented with a bunch of shared pointers so maybe it needs to be a bit fine-tuned for perf (but maybe it's fine). cc @ilia-cher who originally added it

Right, agree, it was added to ThreadLocalState specifically to make it extensible without manually adding Nth field to the struct, we use it to e.g. pass model id and some other custom/specific info that needs to be attached to the model execution (eager thread, TorchScript threads, autograd threads). The impl. need to be re-checked perf-wise, but conceptually it matches well.

**Comment by zou3519** [comment]
Thanks for the info, @dzhulgakov, @ilia-cher! ThreadLocalDebugInfo does exactly what we want here. Agree on the rename and performance checks.

**Comment by ezyang** [comment]
Well, if you look at the code carefully, the "fixed set of slots" is actually done as a linked list. So if you're planning to go wild with the number of context managers, that is still probably a bad idea anyway. Why do you need so many context managers?

In any case, the linked list design for DebugInfoKind means that dynamic registration is relatively simple. Send us a patch to increase the size to a word, and then we can just have people use static pointers to identify entries.

**Comment by zhanyong-wan** [comment]
Thanks for considering, Ed. We need a context manager for each fine-grained knob we provide to the users (e.g. custom XLA flags, execution mode, default precision, excess precision, etc). This is an open set and can keep growing. This seems a principled design to me. Can you clarify why it's probably a bad idea?

Using a void* to identify entries SGTM. Is there any concern with changing the tag from 8-bit to 64-bit?

**Comment by zhanyong-wan** [comment]
Could you assign this issue to me? I don't have the permission to do it. Thanks.

**Comment by zhanyong-wan** [comment]
https://github.com/pytorch/pytorch/pull/180001

**Comment by zhanyong-wan** [comment]
Moved the fix to https://github.com/pytorch/pytorch/pull/180582 .
